week-06-intro-ruby-on-rails/day-01/module-02/readme.md

```diff
@@ -1 +1,115 @@
+
+## Types of Knowledge
+
+* **Declarative Knowledge** > your "what is" knowledge, i.e. describing what something is
+* **Imperative knowledge** > your "how to" knowledge, i.e. describing how to do something
+
+## Parts of A Language
+
+* `Primitives`
+* `Combinations`
+* `Abstractions`
+
+## Discussion Questions
+
+* What is JavaScript? What does it look like?
+* What are some of the primitives in JavaScript?
+  * Think data types, variable declarations, conditionals, functions, etc.
+* How did we use JavaScript to build things? How did we build up from the fundamentals of the language?
+* What could possibly be different in another language? How could we change the syntax, but keep the semantics?
+
+## Base Challenges
+
+*Use `irb` in your terminal for these challenges. Feel free to copy your solutions into a file in Sublime so you can reference them later.*
+
+### Data Types
+
+1. Store your `first_name` in a variable and your `last_name` in another variable.
+
+2. Concatenate your `first_name` and `last_name` variables, and store the output in a new variable called `full_name`.
+
+3. Use <a href="http://ruby-doc.org/core-2.2.0/String.html#method-i-split" target="_blank">`.split`</a> to turn your `full_name` variable into an array.
+
+### Loops
+
+1. Print (`puts`) "Ruby is awesome!" 50 times. Implement this 3 different ways, using:
+  * <a href="http://www.tutorialspoint.com/ruby/ruby_loops.htm" target="_blank">`while`</a>
+  * <a href="http://www.tutorialspoint.com/ruby/ruby_loops.htm" target="_blank">`for`</a>
+  * <a href="http://ruby-doc.org/core-2.0.0/Integer.html#method-i-times" target="_blank">`.times`</a>
+
+2. Save any string to a variable, then create an empty hash called count (`count = {}`). Loop through the string, and count occurrences of each letter. Store the counts in your hash like this example:
+  * For the string `apple`, your `count` hash would look like this: `{a: 1, p: 2, l: 1, e: 1}`.
+
+3. Write a program that gets user input from the terminal and `puts` it until the input is the word `"quit"` or `"q"`.
+  * **Hint:** Use `gets.chomp` instead of `gets` to remove trailing `\n`.
+
+4. Write a program that prints the "bottles of beer on the wall" song:
+
+  ```
+  5 bottles of beer on the wall,
+  5 bottles of beer!
+  Take one down and pass it around,
+  4 bottles of beer on the wall!
+  ...
+  ```
+
+  * Use `gets.chomp` to ask the user how many verses they want to hear.
+  * Make sure your song prints "1 **bottle** of beer".
+  * When the song gets to "0 bottles of beer on the wall", it should print "No more bottles of beer on the wall" instead.
+
+### Iterators: Each
```

Comment:
There are a TON of challenges here. There's no way the majority of the class will get through all of these, and I don't want to stress them out too much since they have the Koans exercises for HW. Maybe move the `each` and `map` challenges to stretch?